python/src/array.cpp

```diff
@@ -131,30 +128,51 @@ PyScalarT validate_shape(T list, const std::vector<int>& shape, int idx) {
   }
 
   if (s == 0) {
-    return pyfloat;
+    return float32;
   }
 
-  PyScalarT type = pybool;
+  Dtype type = bool_;
   for (auto l : list) {
-    PyScalarT t;
+    Dtype t = bool_;
```

Comment:
To avoid needing to change the const of the dtype value, you could revert this to use the PyScalarT and if you get an array, just set it `bool_` (lowest order type) and  let the stack handle the type promotion if you actually get any arrays (which it already will by default). 

Alternatively you could put it in a lambda and add the types to a vector and the reduce the vector with promotion

I think the first option is simpler though.